aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,742 @@
+    formatted_schema = []
+    for row in schema:
+        kwargs = {
+            'name': row['name'],
+            'field_type': row['type'],
+            'mode': row['mode'],
+        }
+
+        if 'fields' in row and row['fields']:
+            kwargs['fields'] = _format_bq_schema_json(row['fields'])
+        formatted_schema.append(bq.SchemaField(**kwargs))
+    return formatted_schema
+
+
+def get_formatted_bq_schema() -> List[bq.SchemaField]:
+    """
+    Get schema for bigquery billing table, as a list of bq.SchemaField objects
+    """
+    return _format_bq_schema_json(get_bq_schema_json())
+
+
+def parse_hail_time(time_str: str) -> datetime:
+    """Parse hail datetime object"""
+    return datetime.strptime(time_str, '%Y-%m-%dT%H:%M:%SZ')
+
+
+def to_bq_time(time: datetime):
+    """Convert datetime to transport datetime for bigquery"""
+    return time.strftime('%Y-%m-%d %H:%M:%S')
+
+
+def get_hail_token():
+    """
+    Get Hail token from local tokens file
+    TODO: look at env var for deploy
+    """
+    if os.getenv('DEV') in ('1', 'true', 'yes'):
+        with open(os.path.expanduser('~/.hail/tokens.json'), encoding='utf-8') as f:
+            config = json.load(f)
+            return config['default']
+
+    assert GCP_PROJECT
+    return read_secret(GCP_PROJECT, 'aggregate-billing-hail-token')
+
+
+def get_hail_credits(entries) -> List[Dict[str, Any]]:
+    """
+    Get a hail credit for each entry
+    """
+
+    hail_credits = [deepcopy(e) for e in entries]
+    for entry in hail_credits:
+
+        entry['topic'] = 'hail'
+        entry['id'] += '-credit'
+        entry['cost'] = -entry['cost']
+        entry['service'] = {
+            'id': 'aggregated-credit',
+            'description': 'Hail credit to correctly account for costs',
+        }
+        entry['sku']['id'] += '-credit'
+        entry['sku']['description'] += '-credit'
+        entry['project'] = HAIL_PROJECT_FIELD
+
+    return hail_credits
+
+
+def get_seqr_credits(entries) -> List[Dict[str, Any]]:
+    """
+    Get a hail credit for each entry
+    """
+
+    seqr_credits = [deepcopy(e) for e in entries]
+    for entry in seqr_credits:
+
+        entry['topic'] = 'seqr'
+        entry['id'] += '-credit'
+        entry['cost'] = -entry['cost']
+        entry['service'] = {
+            'id': 'aggregated-credit',
+            'description': 'Seqr credit to correctly account for costs',
+        }
+        entry['sku']['id'] += '-credit'
+        entry['sku']['description'] += '-credit'
+        entry['project'] = SEQR_PROJECT_FIELD
+
+    return seqr_credits
+
+
+async def get_batches(
+    token: str,
+    billing_project: Optional[str] = None,
+    last_batch_id: Optional[any] = None,
+):
+    """
+    Get list of batches for a billing project with no filtering.
+    (Optional): from last_batch_id
+    """
+
+    qparams = {}
+    if billing_project:
+        qparams['billing_project'] = billing_project
+
+    params = ['q=']
```

Comment:
If we a list billing project, do not insert blank q param

Reply:
Fixed I think, just check my logic here @illusional